_episodes_rmd/05-testthat.Rmd

```diff
@@ -0,0 +1,182 @@
+structure of an R package, we are going to use a helper package to do this for
+us: [`usethis`][usethis].
+
+If you installed and loaded both packages alongside `roxygen2` earlier, you
+should be able to immediately:
+
+```{r use_test, eval=FALSE}
+use_test("center")
+```
+
+A new file should be created called `test-center.R` and pre-filled with a little example. While
+following the explanation of each part, please delete the contents of that example
+test in order to prepare inserting our own. The string within `test_that("…", …)`
+is an explanation of what this test actually tests. In case of our `center()`
+function we expect that "centering works".
+
+Next comes a specific `expect_`ation, usually calling a given function with
+a defined set of arguments (inputs) to check whether that result is `_equal`
+to a known return value (outputs).
+In the function comments, we have already formalised a few of these as examples.
+Copy-and-paste them into the test file, remove the `# should return [1]`
+indicators of the expected results, and wrap the result values into a `c()` to
+enable R to compare the expected values with the output of the two `center()` tests.
+
+```{r center-test, eval=FALSE}
+test_that("centering works", {
+  expect_equal(center(c(1, 2, 3)), c(-1, 0, 1))
+  expect_equal(center(c(1, 2, 3), 1), c(0, 1, 2))
+})
+```
+
+Don't worry that there is no output when you execute either one or both of the
+`expect_equal()` statements, or  the whole `test_that()` block. In good UNIX
+tradition, `testthat` does not bother you with confirming that things are working
+as expected. But there will be plenty of output when some expectations are not
+met, i.e. one or more unit tests fail.
+
+> ## Testing our rescaling functions
+>
+> Apply the above example of converting the examples into unit tests for the
+> `rescale` function as well.
+>
+> > ## Solution
+> > ~~~
+> > test_that("rescaling works", {
+> >   expect_equal(rescale(c(1, 2, 3)), c(0.0, 0.5, 1.0))
+> >   expect_equal(rescale(c(1, 2, 3, 4, 5)), c(0.0, 0.25, 0.5, 0.75, 1.0))
+> > })
+> > ~~~
+> > {: .r}
+> {: .solution}
+{: .challenge}
+
+To conclude this section about creating unit tests, let's again commit our results,
+for example as "Span safety net for TDD".
+
+Finally, have a look at the websites listed on `help(usethis)` to learn about
+how else it can assist you when building packages.
+
+
+## Test-driven development (TDD)
+
+Remember that we updated `rescale()` with lower and upper bounds and default
+values at the end of [episode 2][02]? We had to manually test that change with a
+new example back then. We were repeating ourselves a bit more often than necessary
+back then, weren't we? Of course, there are ways to automate the testing of code
+changes, and to give you quick feedback whether your changes worked, or broke
+anything.
+
+To give ourselves this quick feedback, use `testthat`'s `auto_test_package()`
+and notice the hopefully all green and `OK` `Results`.
```

Comment:
Or use Ctrl-Shift-T in RStudio for running tests. Or do you mention that already? I haven't seen it yet.